File: clients/java/src/main/java/io/dice52/Handshake.java
package io.dice52;

import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.agreement.X25519Agreement;
import org.bouncycastle.crypto.generators.X25519KeyPairGenerator;
import org.bouncycastle.crypto.params.*;
import org.bouncycastle.pqc.crypto.crystals.dilithium.*;
import org.bouncycastle.pqc.crypto.crystals.kyber.*;

import java.security.SecureRandom;

/**
 * Handshake protocol using hybrid KEM (Kyber768 + X25519) and Dilithium3.
 */
public class Handshake {

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    // Kyber768 parameters
    public static final KyberParameters KYBER_PARAMS = KyberParameters.kyber768;

    // Dilithium3 parameters
    public static final DilithiumParameters DILITHIUM_PARAMS = DilithiumParameters.dilithium3;

    private Handshake() {
        // Utility class
    }

    /**
     * Generate a new X25519 key pair.
     * 
     * @return AsymmetricCipherKeyPair containing public and private keys
     */
    public static AsymmetricCipherKeyPair generateX25519Keypair() {
        X25519KeyPairGenerator gen = new X25519KeyPairGenerator();
        gen.init(new X25519KeyGenerationParameters(SECURE_RANDOM));
        return gen.generateKeyPair();
    }

    /**
     * Compute X25519 shared secret.
     */
    public static byte[] x25519SharedSecret(X25519PrivateKeyParameters ourPriv, X25519PublicKeyParameters peerPub) {
        X25519Agreement agreement = new X25519Agreement();
        agreement.init(ourPriv);
        byte[] secret = new byte[agreement.getAgreementSize()];
        agreement.calculateAgreement(peerPub, secret, 0);
        return secret;
    }

    /**
     * Result of hybrid encapsulation.
     */
    public static class HybridEncapsulationResult {
        public final byte[] hybridSharedSecret;
        public final byte[] kyberCiphertext;
        public final byte[] ecdhPub;
        public final X25519PrivateKeyParameters ecdhPriv;

        public HybridEncapsulationResult(byte[] hybridSharedSecret, byte[] kyberCiphertext,
                byte[] ecdhPub, X25519PrivateKeyParameters ecdhPriv) {
            this.hybridSharedSecret = hybridSharedSecret;
            this.kyberCiphertext = kyberCiphertext;
            this.ecdhPub = ecdhPub;
            this.ecdhPriv = ecdhPriv;
        }
    }

    /**
     * Generate a new Kyber768 key pair.
     * 
     * @return AsymmetricCipherKeyPair containing public and private keys
     */
    public static AsymmetricCipherKeyPair generateKemKeypair() {
        KyberKeyPairGenerator gen = new KyberKeyPairGenerator();
        gen.init(new KyberKeyGenerationParameters(SECURE_RANDOM, KYBER_PARAMS));
        return gen.generateKeyPair();
    }

    /**
     * Generate a new Dilithium3 key pair.
     * 
     * @return AsymmetricCipherKeyPair containing public and private keys
     */
    public static AsymmetricCipherKeyPair generateSigningKeypair() {
        DilithiumKeyPairGenerator gen = new DilithiumKeyPairGenerator();
        gen.init(new DilithiumKeyGenerationParameters(SECURE_RANDOM, DILITHIUM_PARAMS));
        return gen.generateKeyPair();
    }

    /**
     * Initiator encapsulate: Alice encapsulates to Bob's public key (Section 7).
     * 
     * @return EncapsulationResult containing shared secret and ciphertext
     */
    public static EncapsulationResult initiatorEncapsulate(KyberPublicKeyParameters peerPub) {
        KyberKEMGenerator kemGen = new KyberKEMGenerator(SECURE_RANDOM);
        org.bouncycastle.crypto.SecretWithEncapsulation encap = kemGen.generateEncapsulated(peerPub);
        return new EncapsulationResult(encap.getSecret(), encap.getEncapsulation());
    }

    /**
     * Responder decapsulate: Bob decapsulates using his private key (Section 7).
     */
    public static byte[] responderDecapsulate(KyberPrivateKeyParameters ourPriv, byte[] ct) {
        KyberKEMExtractor extractor = new KyberKEMExtractor(ourPriv);
        return extractor.extractSecret(ct);
    }

    /**
     * Initiator handshake: hybrid encapsulation to peer's public keys and sign
     * (Section 6.2).
     * 
     * @return HandshakeResult containing message, hybrid shared secret, and ECDH
     *         private key
     */
    public static HandshakeResult initiatorHandshake(
            KyberPublicKeyParameters peerKem,
            X25519PublicKeyParameters peerEcdh,
            DilithiumPrivateKeyParameters idPriv) {
        // Generate ephemeral X25519 key pair
        AsymmetricCipherKeyPair ecdhKeyPair = generateX25519Keypair();
        X25519PublicKeyParameters ecdhPub = (X25519PublicKeyParameters) ecdhKeyPair.getPublic();
        X25519PrivateKeyParameters ecdhPriv = (X25519PrivateKeyParameters) ecdhKeyPair.getPrivate();

        // Kyber encapsulation
        EncapsulationResult encapResult = initiatorEncapsulate(peerKem);

        // X25519 key agreement
        byte[] ssEcdh = x25519SharedSecret(ecdhPriv, peerEcdh);

        // Derive hybrid shared secret
        byte[] ssHybrid = Kdf.deriveHybridSharedSecret(encapResult.sharedSecret, ssEcdh);

        // Sign: context || ciphertext || ecdhPub
        byte[] ecdhPubBytes = ecdhPub.getEncoded();
        byte[] toSign = Kdf.concat(Constants.SIG_CONTEXT, encapResult.ciphertext, ecdhPubBytes);

        DilithiumSigner signer = new DilithiumSigner();
        signer.init(true, idPriv);
        byte[] signature = signer.generateSignature(toSign);

        Types.HandshakeMessage msg = new Types.HandshakeMessage(encapResult.ciphertext, ecdhPubBytes, signature);
        return new HandshakeResult(msg, ssHybrid, ecdhPriv);
    }

    /**
     * Responder handshake: verify signature and hybrid decapsulation (Section 6.3).
     */
    public static byte[] responderHandshake(
            Types.HandshakeMessage msg,
            KyberPrivateKeyParameters kemPriv,
            X25519PrivateKeyParameters ecdhPriv,
            DilithiumPublicKeyParameters peerId)
            throws Dice52Exception.InvalidHandshakeSignature {
        // Verify signature over context || ciphertext || ecdhPub
        byte[] toVerify = Kdf.concat(Constants.SIG_CONTEXT, msg.kyberCt, msg.ecdhPub);

        DilithiumSigner verifier = new DilithiumSigner();
        verifier.init(false, peerId);

        if (!verifier.verifySignature(toVerify, msg.sig)) {
            throw new Dice52Exception.InvalidHandshakeSignature("Handshake signature verification failed");
        }

        // Decapsulate Kyber
        byte[] ssPq = responderDecapsulate(kemPriv, msg.kyberCt);

        // X25519 key agreement
        X25519PublicKeyParameters peerEcdhPub = new X25519PublicKeyParameters(msg.ecdhPub, 0);
        byte[] ssEcdh = x25519SharedSecret(ecdhPriv, peerEcdhPub);

        // Derive hybrid shared secret
        return Kdf.deriveHybridSharedSecret(ssPq, ssEcdh);
    }

    /**
     * Initiator hybrid encapsulate: Alice encapsulates to Bob's public keys.
     */
    public static HybridEncapsulationResult initiatorHybridEncapsulate(
            KyberPublicKeyParameters peerKem,
            X25519PublicKeyParameters peerEcdh) {
        // Generate ephemeral X25519 key pair
        AsymmetricCipherKeyPair ecdhKeyPair = generateX25519Keypair();
        X25519PublicKeyParameters ecdhPub = (X25519PublicKeyParameters) ecdhKeyPair.getPublic();
        X25519PrivateKeyParameters ecdhPriv = (X25519PrivateKeyParameters) ecdhKeyPair.getPrivate();

        // Kyber encapsulation
        EncapsulationResult encapResult = initiatorEncapsulate(peerKem);

        // X25519 key agreement
        byte[] ssEcdh = x25519SharedSecret(ecdhPriv, peerEcdh);

        // Derive hybrid shared secret
        byte[] ssHybrid = Kdf.deriveHybridSharedSecret(encapResult.sharedSecret, ssEcdh);

        return new HybridEncapsulationResult(ssHybrid, encapResult.ciphertext, ecdhPub.getEncoded(), ecdhPriv);
    }

    /**
     * Responder hybrid decapsulate: Bob decapsulates using his private keys.
     */
    public static byte[] responderHybridDecapsulate(
            KyberPrivateKeyParameters kemPriv,
            X25519PrivateKeyParameters ecdhPriv,
            byte[] kyberCt,
            byte[] peerEcdhPub) {
        // Decapsulate Kyber
        byte[] ssPq = responderDecapsulate(kemPriv, kyberCt);

        // X25519 key agreement
        X25519PublicKeyParameters peerPub = new X25519PublicKeyParameters(peerEcdhPub, 0);
        byte[] ssEcdh = x25519SharedSecret(ecdhPriv, peerPub);

        // Derive hybrid shared secret
        return Kdf.deriveHybridSharedSecret(ssPq, ssEcdh);
    }

    /**
     * Result of KEM encapsulation.
     */
    public static class EncapsulationResult {
        public final byte[] sharedSecret;
        public final byte[] ciphertext;

        public EncapsulationResult(byte[] sharedSecret, byte[] ciphertext) {
            this.sharedSecret = sharedSecret;
            this.ciphertext = ciphertext;
        }
    }

    /**
     * Result of handshake initiation.
     */
    public static class HandshakeResult {
        public final Types.HandshakeMessage message;
        public final byte[] sharedSecret;
        public final X25519PrivateKeyParameters ecdhPriv;

        public HandshakeResult(Types.HandshakeMessage message, byte[] sharedSecret,
                X25519PrivateKeyParameters ecdhPriv) {
            this.message = message;
            this.sharedSecret = sharedSecret;
            this.ecdhPriv = ecdhPriv;
        }
    }
}

File: clients/rust/src/handshake.rs
//! Handshake protocol using hybrid KEM (Kyber768 + X25519) and Dilithium3.

use pqcrypto_dilithium::dilithium3;
use pqcrypto_kyber::kyber768;
use pqcrypto_traits::kem::{Ciphertext, SharedSecret};
use pqcrypto_traits::sign::DetachedSignature;
use rand::rngs::OsRng;
use x25519_dalek::{PublicKey as X25519PublicKey, StaticSecret as X25519StaticSecret};

use crate::error::{Dice52Error, Result};
use crate::kdf::derive_hybrid_shared_secret;
use crate::types::{HandshakeMessage, SIG_CONTEXT};

/// Result of hybrid encapsulation
pub struct HybridEncapsulateResult {
    /// Combined hybrid shared secret
    pub ss_hybrid: Vec<u8>,
    /// Kyber ciphertext
    pub kyber_ct: Vec<u8>,
    /// Our X25519 public key
    pub ecdh_pub: Vec<u8>,
    /// Our X25519 private key (for session storage)
    pub ecdh_priv: Vec<u8>,
}

/// Generate a new X25519 key pair
pub fn generate_x25519_keypair() -> (Vec<u8>, Vec<u8>) {
    let secret = X25519StaticSecret::random_from_rng(OsRng);
    let public = X25519PublicKey::from(&secret);
    (public.as_bytes().to_vec(), secret.as_bytes().to_vec())
}

/// Compute X25519 shared secret
pub fn x25519_shared_secret(our_priv: &[u8], peer_pub: &[u8]) -> Result<Vec<u8>> {
    if our_priv.len() != 32 || peer_pub.len() != 32 {
        return Err(Dice52Error::KemError("invalid X25519 key length".into()));
    }

    let mut priv_bytes = [0u8; 32];
    let mut pub_bytes = [0u8; 32];
    priv_bytes.copy_from_slice(our_priv);
    pub_bytes.copy_from_slice(peer_pub);

    let secret = X25519StaticSecret::from(priv_bytes);
    let public = X25519PublicKey::from(pub_bytes);
    let shared = secret.diffie_hellman(&public);

    Ok(shared.as_bytes().to_vec())
}

/// Initiator handshake: hybrid encapsulation to peer's public keys and sign (Section 6.2)
///
/// # Arguments
/// * `peer_kem` - Peer's Kyber public key
/// * `peer_ecdh` - Peer's X25519 public key
/// * `id_priv` - Our Dilithium private key for signing
///
/// # Returns
/// Tuple of (handshake_message, hybrid_shared_secret, our_ecdh_private_key)
pub fn initiator_handshake(
    peer_kem: &kyber768::PublicKey,
    peer_ecdh: &[u8],
    id_priv: &dilithium3::SecretKey,
) -> Result<(HandshakeMessage, Vec<u8>, Vec<u8>)> {
    // Generate ephemeral X25519 key pair
    let (ecdh_pub, ecdh_priv) = generate_x25519_keypair();

    // Kyber encapsulation
    let (ss_pq, ct) = kyber768::encapsulate(peer_kem);

    // X25519 key agreement
    let ss_ecdh = x25519_shared_secret(&ecdh_priv, peer_ecdh)?;

    // Derive hybrid shared secret
    let ss_hybrid = derive_hybrid_shared_secret(ss_pq.as_bytes(), &ss_ecdh);

    // Sign: context || ciphertext || ecdh_pub
    let ct_bytes = ct.as_bytes();
    let mut to_sign = SIG_CONTEXT.to_vec();
    to_sign.extend_from_slice(ct_bytes);
    to_sign.extend_from_slice(&ecdh_pub);

    let sig = dilithium3::detached_sign(&to_sign, id_priv);

    Ok((
        HandshakeMessage {
            kyber_ct: ct_bytes.to_vec(),
            ecdh_pub,
            sig: sig.as_bytes().to_vec(),
        },
        ss_hybrid.to_vec(),
        ecdh_priv,
    ))
}

/// Responder handshake: verify signature and hybrid decapsulation (Section 6.3)
///
/// # Arguments
/// * `msg` - Received handshake message
/// * `kem_priv` - Our Kyber private key
/// * `ecdh_priv` - Our X25519 private key
/// * `peer_id` - Peer's Dilithium public key for verification
///
/// # Returns
/// Hybrid shared secret
pub fn responder_handshake(
    msg: &HandshakeMessage,
    kem_priv: &kyber768::SecretKey,
    ecdh_priv: &[u8],
    peer_id: &dilithium3::PublicKey,
) -> Result<Vec<u8>> {
    // Verify signature over context || ciphertext || ecdh_pub
    let mut to_verify = SIG_CONTEXT.to_vec();
    to_verify.extend_from_slice(&msg.kyber_ct);
    to_verify.extend_from_slice(&msg.ecdh_pub);

    let sig = dilithium3::DetachedSignature::from_bytes(&msg.sig)
        .map_err(|_| Dice52Error::InvalidHandshakeSignature)?;

    dilithium3::verify_detached_signature(&sig, &to_verify, peer_id)
        .map_err(|_| Dice52Error::InvalidHandshakeSignature)?;

    // Decapsulate Kyber
    let ct = kyber768::Ciphertext::from_bytes(&msg.kyber_ct)
        .map_err(|_| Dice52Error::KemError("invalid ciphertext".into()))?;

    let ss_pq = kyber768::decapsulate(&ct, kem_priv);

    // X25519 key agreement
    let ss_ecdh = x25519_shared_secret(ecdh_priv, &msg.ecdh_pub)?;

    // Derive hybrid shared secret
    let ss_hybrid = derive_hybrid_shared_secret(ss_pq.as_bytes(), &ss_ecdh);

    Ok(ss_hybrid.to_vec())
}

/// Initiator hybrid encapsulate: Alice encapsulates to Bob's public keys
///
/// # Arguments
/// * `peer_kem` - Peer's Kyber public key
/// * `peer_ecdh` - Peer's X25519 public key
///
/// # Returns
/// HybridEncapsulateResult containing hybrid shared secret, ciphertexts, and keys
pub fn initiator_hybrid_encapsulate(
    peer_kem: &kyber768::PublicKey,
    peer_ecdh: &[u8],
) -> Result<HybridEncapsulateResult> {
    // Generate ephemeral X25519 key pair
    let (ecdh_pub, ecdh_priv) = generate_x25519_keypair();

    // Kyber encapsulation
    let (ss_pq, ct) = kyber768::encapsulate(peer_kem);

    // X25519 key agreement
    let ss_ecdh = x25519_shared_secret(&ecdh_priv, peer_ecdh)?;

    // Derive hybrid shared secret
    let ss_hybrid = derive_hybrid_shared_secret(ss_pq.as_bytes(), &ss_ecdh);

    Ok(HybridEncapsulateResult {
        ss_hybrid: ss_hybrid.to_vec(),
        kyber_ct: ct.as_bytes().to_vec(),
        ecdh_pub,
        ecdh_priv,
    })
}

/// Responder hybrid decapsulate: Bob decapsulates using his private keys
///
/// # Arguments
/// * `kem_priv` - Our Kyber private key
/// * `ecdh_priv` - Our X25519 private key
/// * `kyber_ct` - Received Kyber ciphertext
/// * `peer_ecdh_pub` - Peer's X25519 public key
///
/// # Returns
/// Hybrid shared secret
pub fn responder_hybrid_decapsulate(
    kem_priv: &kyber768::SecretKey,
    ecdh_priv: &[u8],
    kyber_ct: &[u8],
    peer_ecdh_pub: &[u8],
) -> Result<Vec<u8>> {
    // Decapsulate Kyber
    let ct = kyber768::Ciphertext::from_bytes(kyber_ct)
        .map_err(|_| Dice52Error::KemError("invalid ciphertext".into()))?;

    let ss_pq = kyber768::decapsulate(&ct, kem_priv);

    // X25519 key agreement
    let ss_ecdh = x25519_shared_secret(ecdh_priv, peer_ecdh_pub)?;

    // Derive hybrid shared secret
    let ss_hybrid = derive_hybrid_shared_secret(ss_pq.as_bytes(), &ss_ecdh);

    Ok(ss_hybrid.to_vec())
}

// Legacy functions for backwards compatibility (Kyber-only)

/// Initiator encapsulate: Alice encapsulates to Bob's public key (Kyber-only, legacy)
///
/// # Arguments
/// * `peer_pub` - Peer's Kyber public key
///
/// # Returns
/// Tuple of (shared_secret, ciphertext)
pub fn initiator_encapsulate(peer_pub: &kyber768::PublicKey) -> (Vec<u8>, Vec<u8>) {
    let (ss, ct) = kyber768::encapsulate(peer_pub);
    (ss.as_bytes().to_vec(), ct.as_bytes().to_vec())
}

/// Responder decapsulate: Bob decapsulates using his private key (Kyber-only, legacy)
///
/// # Arguments
/// * `our_priv` - Our Kyber private key
/// * `ct` - Received ciphertext
///
/// # Returns
/// Shared secret
pub fn responder_decapsulate(our_priv: &kyber768::SecretKey, ct: &[u8]) -> Result<Vec<u8>> {
    let ct = kyber768::Ciphertext::from_bytes(ct)
        .map_err(|_| Dice52Error::KemError("invalid ciphertext".into()))?;

    let ss = kyber768::decapsulate(&ct, our_priv);
    Ok(ss.as_bytes().to_vec())
}

/// Generate a new Kyber768 key pair
pub fn generate_kem_keypair() -> (kyber768::PublicKey, kyber768::SecretKey) {
    kyber768::keypair()
}

/// Generate a new Dilithium3 key pair
pub fn generate_signing_keypair() -> (dilithium3::PublicKey, dilithium3::SecretKey) {
    dilithium3::keypair()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_x25519_key_exchange() {
        let (pub_a, priv_a) = generate_x25519_keypair();
        let (pub_b, priv_b) = generate_x25519_keypair();

        let ss_a = x25519_shared_secret(&priv_a, &pub_b).unwrap();
        let ss_b = x25519_shared_secret(&priv_b, &pub_a).unwrap();

        assert_eq!(ss_a, ss_b);
    }

    #[test]
    fn test_encapsulate_decapsulate() {
        let (pub_key, priv_key) = generate_kem_keypair();
        let (ss1, ct) = initiator_encapsulate(&pub_key);
        let ss2 = responder_decapsulate(&priv_key, &ct).unwrap();
        assert_eq!(ss1, ss2);
    }

    #[test]
    fn test_hybrid_encapsulate_decapsulate() {
        let (kem_pub, kem_priv) = generate_kem_keypair();
        let (ecdh_pub, ecdh_priv) = generate_x25519_keypair();

        let result = initiator_hybrid_encapsulate(&kem_pub, &ecdh_pub).unwrap();

        let ss2 =
            responder_hybrid_decapsulate(&kem_priv, &ecdh_priv, &result.kyber_ct, &result.ecdh_pub)
                .unwrap();

        assert_eq!(result.ss_hybrid, ss2);
    }

    #[test]
    fn test_hybrid_handshake() {
        let (_kem_pub_a, _kem_priv_a) = generate_kem_keypair();
        let (kem_pub_b, kem_priv_b) = generate_kem_keypair();
        let (_ecdh_pub_a, _ecdh_priv_a) = generate_x25519_keypair();
        let (ecdh_pub_b, ecdh_priv_b) = generate_x25519_keypair();
        let (id_pub_a, id_priv_a) = generate_signing_keypair();

        // Alice initiates hybrid handshake to Bob
        let (msg, ss_alice, _ecdh_priv) =
            initiator_handshake(&kem_pub_b, &ecdh_pub_b, &id_priv_a).unwrap();

        // Bob responds using hybrid decapsulation
        let ss_bob = responder_handshake(&msg, &kem_priv_b, &ecdh_priv_b, &id_pub_a).unwrap();

        assert_eq!(ss_alice, ss_bob);
    }
}
